R/ReportDocument.R

```diff
@@ -58,6 +58,71 @@ c.ReportDocument <- function(...) {
   out
 }
 
+#' @export
+label <- function(object) UseMethod("label", object)
+
+#' @export
+label.ReportDocument <- function(object) {
+  label <- attr(object, which = "label", exact = TRUE)
+  label %||% character(0L)
+}
+
+#' @export
+label.ReportCard <- function(object) {
+  # TODO: soft deprecate
+  object$get_name()
+}
+
+
+#' @export
+`label<-` <- function(object, value) UseMethod("label<-", object)
+
+#' @export
+`label<-.ReportDocument` <- function(object, value) {
+  attr(object, which = "label") <- value
+  object
+}
+
+#' @export
+`label<-.ReportCard` <- function(object, value) {
+  object$set_name(value)
+  object
+}
+
+#' @export
+id <- function(object) UseMethod("id", object)
+
+#' @export
+id.ReportDocument <- function(object) {
+  id <- attr(object, which = "id", exact = TRUE)
+  id %||% character(0L)
+}
+
+#' @export
+id.ReportCard <- function(object) {
+  object$get_id()
+}
+
+#' @export
+`id<-` <- function(object, value) UseMethod("id<-", object)
+
+#' @export
+`id<-.ReportDocument` <- function(object, value) {
+  if (identical(id(object), character(0L))) {
+    attr(object, which = "id") <- value
+    return(object)
+  }
+  warning("'id' for reporter is already set, skipping...")
+  object
+}
+
+#' @export
+`id<-.ReportCard` <- function(object, ...) {
+  # Value is not needed as object can generate id
+  object$generate_id()
+  object
+}
+
```

Comment:
I'm not sure about keeping these 2 sets of functions (`id()`, `id() <-`, `label()` and `label() <-`) exported in `teal.reporter`

Pros:

- Relieve the code of some complexity as they add an API to handle the getters and setters without having `attr()` repeated call to access the attributes directly.
- Allow for better support for both `ReportDocument` and `ReportCard`

Cons:

- Not necessary outside of `teal.reporter` with 1 exception in `teal` (with a workaround possible: using `attr() <-` directly.

Alternatively, we could only register the methods and not export the generic (that is, remove the `@export` hint from the generic)